cmd/crank/completion/completion.go

```diff
@@ -0,0 +1,259 @@
+// Package completion contains Crossplane CLI completions.
+package completion
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/posener/complete"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/discovery/cached/memory"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	errFmtResourceTypeNotFound = "the server doesn't have a resource type %q"
+)
+
+// Predictors returns all supported predictors.
+func Predictors() map[string]complete.Predictor {
+	yamlPredictor := complete.PredictOr(
+		complete.PredictFiles("*.yml"),
+		complete.PredictFiles("*.yaml"),
+	)
+	return map[string]complete.Predictor{
+		"file":                   complete.PredictFiles("*"),
+		"xpkg_file":              complete.PredictFiles("*.xpkg"),
+		"yaml_file":              yamlPredictor,
+		"directory":              complete.PredictDirs("*"),
+		"file_or_directory":      complete.PredictOr(complete.PredictFiles("*"), complete.PredictDirs("*")),
+		"yaml_file_or_directory": complete.PredictOr(yamlPredictor, complete.PredictDirs("*")),
+		"namespace":              namespacePredictor(),
+		"context":                contextPredictor(),
+		"k8s_resource":           kubernetesResourcePredictor(),
+		"k8s_resource_name":      kubernetesResourceNamePredictor(),
+	}
+}
+
+func kubernetesResourcePredictor() complete.PredictFunc {
```

Comment:
i think this file could use a bit more comments to help the reader understand the key parts of the implementation and orient themselves more easily. what do you think?

Reply:
Fair point, I've added a couple of comments to the code.